certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -127,6 +128,8 @@ def mod_ssl_conf_src(self):
         config_filename = "options-ssl-nginx.conf"
         if self.version < (1, 5, 9):
             config_filename = "options-ssl-nginx-old.conf"
+        elif self.version < (1, 13, 0) or ssl.OPENSSL_VERSION_INFO < (1, 1, 1):
```

Comment:
Looking at the output of `nginx -V` actually makes me suspect that it may not in fact be good enough: it says "built with OpenSSL 1.1.1  11 Sep 2018" which reads to me as if it's vendorized into the nginx package. In which case the results of `ssl.OPENSSL_VERSION_INFO` might be inaccurate. Investigating this now.

Reply:
As for 2, it prints a warning to stderr: `nginx: [warn] invalid value "TLSv1.3" in [path]:[line]`

Reply:
So we could do that. I don't love doing it that way; that would take away some of the nice abstraction where it's currently all neatly bundled in a property.

Reply:
Ok great news, looks like nginx -V is just running `OpenSSL_version` under the hood. https://github.com/nginx/nginx/blob/4bf4650f2f10f7bbacfe7a33da744f18951d416d/src/core/nginx.c#L430

Reply:
So anyway, looks like the version info has been there since 1.17.2. https://github.com/nginx/nginx/commit/fd4da8922851cf5b7ada71ac18d54c9fe5be520a#diff-0270f841825e1c3a228f4800b9ab71cd

Reply:
So for Nginx between 1.13.0 and 1.17.1, we wouldn't be able to check OpenSSL version number.

Reply:
If anything, I think I'd prefer to shell out to openssl to check its version directly. Doesn't look like Nginx is doing any vendorizing, and we're not accessing the snap version of Nginx.

Reply:
> So we could do that. I don't love doing it that way; that would take away some of the nice abstraction where it's currently all neatly bundled in a property.

I agree the code is nice but I think it's not actually checking the right thing. It might be a good enough approximation, but if we stick with the code as is, I think we should make sure we're OK with our behavior when we're wrong about the OpenSSL version. I think it'll probably happen to people eventually.

> As for 2, it prints a warning to stderr: nginx: [warn] invalid value "TLSv1.3" in [path]:[line]

What are the chances it just ignores the value "TLSv1.3" in the list of protocols rather than ignoring the list entirely and falling back to some (probably outdated) default? I think it'd be best to try to actually figure out the version of OpenSSL nginx is using to avoid these warning messages, but if that's hard, it may give us an easy way to get the behavior we want.

> If anything, I think I'd prefer to shell out to openssl to check its version directly. Doesn't look like Nginx is doing any vendorizing, and we're not accessing the snap version of Nginx.

I think there's a slight difference in shelling out to the openssl command line vs. calling a function in the library you're linked against. I also don't think we can count on the openssl command line always being available either. nginx just needs the library.

Reply:
A couple tests:

```
nginx version: nginx/1.15.5 (yes TLSv1.3)
built with OpenSSL 1.0.2g  1 Mar 2016 (no TLSv1.3)
ssl_protocols TLSv1.2 TLSv1.3;
no warning printed
connects perfectly fine, using TLSv1.2.

nginx version: nginx/1.15.5 (yes TLSv1.3)
built with OpenSSL 1.0.2g  1 Mar 2016 (no TLSv1.3)
ssl_protocols TLSv1.3;
no warning printed
fails to load https site
loads http site
```

But look here:
https://github.com/nginx/nginx/blob/c17bc31d41a0372002115899a2c64e89aeca7e7d/src/event/ngx_event_openssl.c#L344-L349

Ok so here's what happens. Nginx understands the TLS1.3 flag, so it parses it fine. But the way it communicates SSL versions to OpenSSL is if it DOESN'T see a flag, it tells OpenSSL NOT to use that SSL version... all of which is conditioned on the OpenSSL flag being defined in the first place. So including TLS1.3 is not a problem, because Nginx won't try to communicate anything in the first place.

And further, this is how it was [added in the first place](https://github.com/nginx/nginx/commit/9a37eb3a62130473596e0e4c2e388d80bdb14956), so this is true for all versions of Nginx that support TLS1.3.

What this tells me is that actually there's really no need to bother checking OpenSSL version in the first place, we can always just add TLSv1.3 if Nginx understands it.

Reply:
Nice! I saw the same behavior here.